Codechange: introduce a framework for all our timers

TrueBrain opened the pull request:
## Motivation / Problem

We have a few things we would like to address:
- Autosaves based on real-time, not in-game time
- Similar for LinkGraphs

Additionally, we would like to split up game-time in "calendar" and "economy", to allow speeding up one and not the other.
This is all a bit tricky in the current code-base, as we track these kind of things all over the place.

So, I was like: can't we make a single system for Interval and Timeout timers, where you can just say based on what "clock" you want it to work.

This PR is a result of that idea. It just puts down the foundation, and converts game-date and window-based timers.

## Description

`IntervalTimer` and `TimeoutTimer` use RAII, and can be used to replace all the time-based timeouts, lag-detection, "execute every N" we have.
As it uses RAII, you can safely use it as static variable, class member, temporary variable, etc. As soon as it goes out-of-scope, it will be safely removed.
This allows for much easier to read code when it comes to intervals.

This replaces `GUITimer` for Windows.

Additional, for all Window related function I replaced `MILLISECONDS_PER_TICK` to `30`, as that is actual the intent. The Window system is completely uncoupled from the game-ticks, so it shouldn't make the suggestion it is different.

This PR is just the start of things. It got a bit bigger than I would like, but I wanted to show that we can use this to replace all oddities we have in our code. As the Window system has the most oddities, I replaced all time-related events there with this new system. Sometimes it can be a bit hard to see if the work I did is correct, as some code is just plain weird. I did my best.
Additionally, to make sure we can convert our game-date related timers too, tackled those. Found some fun issues with them, and addressed those.

Future work would include:
- Replace all lag-related network detection in a similar way.
- Move autosave / linkgraph to real-time, instead of game-time.
And I am sure I miss many more.

## Limitations / remarks

### O(n)

Every timer registered is checked on every progression of time. (read: `O(n)`). This is strictly seen not needed, and we could use a priority-tree, to check if only the top entry progressed sufficiently.
But, to not overcomplicate a relative few entries on a list, I didn't go that route. This PR puts down the foundation; we can always extend with this at a later time, within the interface of the current implementation.

### Templating

This PR makes strong use of templating; in theory we could split every time up in their own function, and do away with the templating. I mostly went for this route, to make it easier to switch from one time to the other, and to make sure all timers look the same, no matter what their trigger is.

### Magic numbers (resolved)

Many places use a `std::chrono`, instead of an enum for their timings. This was already there, or implied, but in theory it should be put in an enum. This PR only was already long enough, that I decided not to do that in this PR.
EDIT: after talking this over, decided to remove all enum references instead. This makes more sense, as this keeps the information local. The documentation of the enums were .. non-descriptive at best. This way the context is clear instantly.

### Related bugs

When doing a PR like this, you can't help but finding all kind of weird bugs. Like the setting `errmsg_duration` is in 3-second intervals, not 1-second what the tooltip suggests. Framerate graphs jump all over the place.
To keep a focus, I pushed those all out of scope. This PR should be as broken on these subjects as it was before this PR.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

LordAro commented on src/company_gui.cpp:
```diff
@@ -2674,11 +2679,10 @@ struct CompanyWindow : Window
-	{
-		/* redraw the window every now and then */
+	/** Redraw the window on a regular interval. */
+	IntervalTimer<TimerWindow> redraw_interval = {std::chrono::seconds(3), [this](auto) {
```
Does the actual time value ever vary? Is it likely to? Perhaps it could be a template parameter instead?

PeterN replied:
There is at least one case where the timer is scaled by scaled font size.

TrueBrain replied:
There is at least one instance of `std::chrono::seconds(3 * _settings_client.gui.errmsg_duration)`, and I suspect there will be a few more over time.

So yeah, sadly, no template parameter :(

TrueBrain committed:
Codechange: migrate all game-time-related timers to the new framework

TrueBrain committed:
Codechange: introduce a framework for all our timers

IntervalTimer and TimeoutTimer use RAII, and can be used to replace
all the time-based timeouts, lag-detection, "execute every N" we
have.
As it uses RAII, you can safely use it as static variable, class
member, temporary variable, etc. As soon as it goes out-of-scope,
it will be safely removed.
This allows for much easier to read code when it comes to intervals.

TrueBrain committed:
Codechange: migrate all Window-related timers to the new framework

This means we also say goodbye to GUITimers.

rubidium42 commented on src/engine.cpp:
```diff
@@ -1110,6 +1112,11 @@ void EnginesMonthlyLoop()
 	}
 }
 
+static IntervalTimer<TimerGameCalendar> _engines_monthly({TimerGameCalendar::MONTH, TimerGameCalendar::Priority::ENGINE}, [](auto)
```
Shouldn't this line practically replace the `EnginesMonthlyLoop()` like it does for `EnginesDailyLoop()`?

TrueBrain replied:
Yes, except for some reason this function is also called from the cheats menu.

PS: similar with `ResetSignalVariant` btw.

frosch123 commented on src/timer/timer_game_calendar.h:
```diff
@@ -0,0 +1,76 @@
+ *
+ * Additionally, you need to set a priority. To ensure deterministic behaviour, events are executed
+ * in priority. It is important that if you assign NONE, you do not use Random() in your callback.
+ * Other than that, make sure you only set one callback per priority.
```
This is not specific to Random().
All non-GUI-code handlers should be sorted, so all priorities should have a single handler only.
Currently Priority::ENGINE is used twice, and I can imagine that already desyncs. Engine and track-types become available depending on each other, ...

TrueBrain replied:
They aren't used twice. Well, strictly seen, but one time on Day, and one time on Month. That order is well defined. So it is deterministic currently, and I wouldn't see where any desync would be possible currently :)

Did you overlook that they are on different triggers, or am I missing something?

frosch123 replied:
Ah yes, I missed that Day/Month thing. Can there be an assertion in the gameloop timer, that the handlers with priority are strictly ordered?

TrueBrain replied:
can do!

I also planned to do it, but forgot .. so tnx for pointing out it was missing :D

As if this fails means that a developer made a big boo-boo, the code only runs for assert-enabled builds. Means I can just `assert` to detect issues :)

TrueBrain committed:
Codechange: validate the developer didn't schedule two timers on the same trigger/priority
